docs/content/en/docs/Coding Milestones/PoC2023q1/outline.md

```diff
@@ -122,6 +122,34 @@ what should be returned as well as implementing it.  This will
 certainly affect the syncer between mailbox workspace and edge
 cluster, and the summarization part will affect the status summarizer.
 
+### Good handling of workload conflicts
+
+We could start by handling workload conflicts in a very simple way:
+treating each as an error.  Later development can handle them better,
+as outlined [later](#edgeplacement-objects).
+
+### Denaturing/renaturing
+
+We could start by not doing this.  For some resources, the effect of
+leaving these resources natured in the center is only to add
+authorizations in the center that are not needed and are undesired in
+a well-secured environment but tolerable in early demonstrations ---
+provided that there is not a conflict with an object of the same name
+that is positively desired in the center.  In particular, these are:
+`ClusterRole`, `ClusterRoleBinding`, `Role`, `RoleBinding`,
+`ServiceAccount`.
+
+For some kinds of object, lack of denaturing/renaturing means that
+edge-mc will simply not be able to support workloads that contain such
+objects.  These are: `MutatingWebhookConfiguration`,
+`ValidatingWebhookConfiguration`, `LimitRange`, `ResourceQuota`.
+
+For some resources, the need to denature is only a matter of
+anticipation.  `FlowSchema` and `PriorityLevelConfiguration` currently
+are not interpreted by kcp servers and so are effectively already
+denatured there.  Hopefully they will be given interpretations in the
+future, and then those resources will join the previous category.
```

Comment:
This is really important feature for edge syncer to support these resources. 
- replace to some alternative kind in workspace side
- edge syncer pull the alternative kind and replace it with native one. 
I think this should be one of the priority item.

Reply:
Denaturing and renaturing are coupled.  The denaturing is done by a view, the renaturing is done by (the placement translator while the TMC syncer is used, the EMC syncer once it is developed).

I see now that #242 does not yet clearly enough ask the relevant questions here.  So let me ask them here.  First I should note that I hope that kcp will be rebased onto Kubernetes release 1.26 soon, which will make https://kubernetes.io/docs/reference/kubernetes-api/extend-resources/validating-admission-policy-v1alpha1/ usable; I wonder whether (as suggested at the end of that review PDF you shared) C2P could use `ValidatingAdmissionPolicy` objects.  In that future, does C2P need `MutatingWebhookConfiguration`, `ValidatingWebhookConfiguration`, `LimitRange`, or `ResourceQuota`?  More precisely, could we demo an interesting subset of C2P without using those resources?

Reply:
> could we demo an interesting subset of C2P without using those resources?

Kyverno does not require syncing those resources for C2P demo. `MutatingWebhookConfiguration` and/or `ValidatingWebhookConfiguration` are generated after deploying Kyverno runtime on p-cluster.

Reply:
@yuji-watanabe-jp : thank you, I just want to follow up to make sure I understand correctly.  When the Kyverno controller generates a `MutatingWebhookConfiguration` or `ValidatingWebhookConfiguration` object in the edge cluster, we have to be careful about which apiserver this object is written to.  It needs to be written to the apiserver that it is intended to affect.  My concern is what happens while using the TMC syncer.  If the webhook configuration object is written to the kcp workspace but is intended to configure calls to a container running in the edge cluster then this will be a broken configuration.  Is the Kyverno controller described by a Deployment object that will be downsynced with the TMC syncer rewiring its apiserver client config to point back to the mailbox workspace?

Reply:
A Kubernetes `Service` does not create webhook configuration objects, but a `Pod` (created as a consequence of a `Deployment`) might.

When TMC downsyncs a `Deployment` object, the syncer modifies the `PodTemplaceSpec` in that `Deployment` so that as it appears on the pcluster, the `Pod` objects that get created will --- IF they use the regular way of connecting --- connect to the kcp workspace rather than the pcluster's apiserver.

Reply:
Sorry, I am just using service in general sense, but i do not have much concern on this. When kyverno starts on p-cluster, kyverno configure webhook on the cluster the kyverno is running unless any unexpected mutation happens in the manifest. It is p-cluster. nothing related to kcp workspace anymore.

Reply:
if the syncer is doing something bad when syncing resources from workspace to p-cluster, something happens, but I have not seen the situation so far in our experiments.

Reply:
@yuji-watanabe-jp , are you expecting the manifest mutation that the TMC syncer normally does?  They do not consider it "bad", it is the intended behavior of that thing.

Reply:
OK, "bad" might be wrong word. I meant unexpected for me. If it is expected from TMC perspective, we need to deal with it and maybe need to find workaround. If we use edge-syncer, it is different story. so we can revisit this later.

Reply:
@yuji-watanabe-jp I think it's important to clarify how Kyverno gets installed on the p-Cluster. Based on how it gets installed the behavior of Kyverno controllers can be very different.  If you install Kyverno directly on the pCluster, without any involvement of kcp, then the Kyverno controllers know nothing of kcp and only talk to the  pCluster API server. `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` are created in the pCluster by the Kyverno controllers.  But ... if you install kyverno from on a kcp workspace, and you have TMC syncer configured to sync the Kyverno deployment to the pCluster, then the TMC syncer mutators update the deployment to point back to the kcp Workspace. In this scenario, the Kyverno CRDs (and other auxiliary resources such as secrets etc) are installed in the kcp workspace. The Kyverno controllers will then try to create `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` on the Kcp workspace, and also it may not be able to build properly a policy report for the pCluster as they are not watching/listing objects on the pCluster but on the kcp workspace. 

I remember some time ago you showed a chart where you were installing Kyverno with OLM and just syncing the OLM templates. In that case, you may have the benefit of installing on the pCluster without being affected by the syncer mutators (which currently only manipulate Deployments, Pods and ReplicaSets). Is that still the case ?

Reply:
I think we have some choices here.  Perhaps installing Kyverno via OLM is the better choice in the near term.  In that case, we would have the issue TMC upsync of the policy reports from the edge cluster to the mailbox workspace.

Reply:
@pdettori you are right. we have been deploying Kyverno to p-cluster with TMC syncer + OLM. We confirmed that works. 
Without OLM, we need to see how to install cluster-scope (ClusterRoleBinding and ClusterRole) via TMC syncer. 
So, we can keep the choice with OLM for installation via TMC syncer. The issue of TMC upsync of the policy reports still remains.

Edge syncer will do not apply such mutation, and so this will never happen there, and resolve the issue of upsync.

Reply:
@yuji-watanabe-jp Cluster-scope deployment and the webhook issue are different. Just to emphasis (and maybe close :-)  ) this discussion - if you plan to deploy  Kyverno  through KCP (i.e., into a KCP workspace) than due to the mutating that TMC performs on the deployed pods your webhooks probably won't work correctly. If you can (for the first stage) deploy the controllers directly on the pCluster(s) - than you avoid the TMC mutating and your pCluster webhooks will probably work without an issue.

Reply:
The upsync seems not related to this , as for both cases (deploy throgh KCP and externally to KCP) , we can do the same: (1) define the criteria of what to upsync through KCP (2) A controller on the pCLuster (part of the Kyverno  syncer) will need to support this upSync (e.g., set the labels/annotations, etc..)

Reply:
Is the controller that creates the policy objects something that we control? Can we make it put needed labels and annotations on the policy objects, at least in the TMC interim?

Reply:
we control it. we can add them.